spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/reactive/PubSubReactiveFactory.java

```diff
@@ -77,116 +86,62 @@ public PubSubReactiveFactory(PubSubSubscriberOperations subscriberOperations, Sc
 	public Flux<AcknowledgeablePubsubMessage> poll(String subscriptionName, long pollingPeriodMs) {
 
 		return Flux.create(sink -> {
-
-			Scheduler.Worker subscriptionWorker = this.scheduler.createWorker();
-
 			sink.onRequest((numRequested) -> {
 				if (numRequested == Long.MAX_VALUE) {
-					// unlimited demand
-					subscriptionWorker.schedulePeriodically(
-							new NonBlockingUnlimitedDemandPullTask(subscriptionName, sink), 0, pollingPeriodMs, TimeUnit.MILLISECONDS);
+					pollingPull(subscriptionName, pollingPeriodMs, sink);
 				}
 				else {
-					subscriptionWorker.schedule(new BlockingLimitedDemandPullTask(subscriptionName, numRequested, sink));
+					backpressurePull(subscriptionName, numRequested, sink);
 				}
 			});
-
-			sink.onDispose(subscriptionWorker);
-
 		});
 	}
 
-	private abstract class PubSubPullTask implements Runnable {
-
-		protected final String subscriptionName;
-
-		protected final FluxSink<AcknowledgeablePubsubMessage> sink;
-
-		PubSubPullTask(String subscriptionName, FluxSink<AcknowledgeablePubsubMessage> sink) {
-			this.subscriptionName = subscriptionName;
-			this.sink = sink;
-		}
-
-		/**
-		 * Retrieve up to a specified number of messages, sending them to the subscription.
-		 * @param demand maximum number of messages to retrieve
-		 * @param block whether to wait for the first message to become available
-		 * @return number of messages retrieved
-		 */
-		protected int pullToSink(int demand, boolean block) {
-
-			List<AcknowledgeablePubsubMessage> messages =
-					PubSubReactiveFactory.this.subscriberOperations.pull(this.subscriptionName, demand, !block);
-
-			if (!this.sink.isCancelled()) {
-				messages.forEach(sink::next);
-			}
-
-			return messages.size();
-		}
+	private void pollingPull(String subscriptionName, long pollingPeriodMs,
+			FluxSink<AcknowledgeablePubsubMessage> sink) {
+		Disposable disposable = Flux
+				.interval(Duration.ZERO, Duration.ofMillis(pollingPeriodMs), scheduler)
+				.flatMap(ignore -> pullAll(subscriptionName))
+				.subscribe(sink::next, sink::error);
 
+		sink.onDispose(disposable);
 	}
 
-	/**
-	 * Runnable task issuing blocking Pub/Sub Pull requests until the specified number of
-	 * messages has been retrieved.
-	 */
-	private class BlockingLimitedDemandPullTask extends PubSubPullTask {
+	private Flux<AcknowledgeablePubsubMessage> pullAll(String subscriptionName) {
+		CompletableFuture<List<AcknowledgeablePubsubMessage>> pullResponseFuture = this.subscriberOperations
+				.pullAsync(subscriptionName, Integer.MAX_VALUE, true).completable();
```

Comment:
side note: use `true` for the third argument, defines the `return_immediately` as `true`. However according to the Google PubSub API docs, this is not recommended and the argument is actually deprecated (see https://cloud.google.com/pubsub/docs/reference/rpc/google.pubsub.v1#pullrequest). 

What to do with it? Separate bug?

Reply:
Thank you for catching this; we actually did not realize this field recently got deprecated.
Let's keep the behavior consistent with the past in this PR.

#2247 to track.